src/server_manager/web_app/ui_components/outline-per-key-data-limit-dialog.ts

```diff
@@ -0,0 +1,341 @@
+            color: rgb(123, 123, 123);
+          }
+        }
+      </style>
+      <paper-dialog id="container">
+        <div id="headerSection">
+          <iron-icon id="dataLimitIcon" icon="icons:perm-data-setting"></iron-icon>
+          <h3>${this.localize('per-key-data-limit-dialog-title', 'keyName', keyName)}</h3>
+        </div>
+        <div id="menuSection">
+          <paper-checkbox ?checked=${this.showMenu} @tap=${this._setCustomLimitTapped}>
+            ${this.localize('per-key-data-limit-dialog-set-custom')}
+          </paper-checkbox>
+          <div id="menu" ?hidden=${!this.showMenu}>
+            <paper-input
+              id="dataLimitInput"
+              label=${this.localize('data-limit')}
+              always-float-label
+              allowed-pattern="[0-9\\.]"
+              pattern="[0-9]+(\\.[0-9]*)?"
+              auto-validate
+              value=${this._initialValue()}
+              size="7"
+              @keyup=${this._setSaveButtonDisabledState}
+            >
+            </paper-input>
+            <paper-dropdown-menu id="unitsDropdown" noink>
+              <paper-listbox
+                id="unitsListbox"
+                slot="dropdown-content"
+                attr-for-selected="name"
+                selected="${this._initialUnit()}"
+              >
+                <paper-item name="GB">GB</paper-item>
+                <paper-item name="MB">MB</paper-item>
+              </paper-listbox>
+            </paper-dropdown-menu>
+          </div>
+        </div>
+        <div id="buttonsSection">
+          <paper-button id="save" ?disabled=${!this.enableSave} @tap=${this._sendSaveEvent}>${
+        this.localize('save')}</paper-button>
+          <paper-button @tap=${this.close}>${this.localize('cancel')}</paper-button>
+        </div>
+      </paper-dialog>
+    `;
+  }
+
+  private _queryAs<T extends HTMLElement>(selector: string): T {
+    return this.shadowRoot.querySelector(selector) as T;
+  }
+
+  private get _input(): PaperInputElement {
+    return this._queryAs<PaperInputElement>('#dataLimitInput');
+  }
+
+  private _dataLimitValue() {
+    return Number(this._input.value);
+  }
+
+  private _dataLimitType() {
+    return this._queryAs<PaperDropdownMenuElement>('#unitsDropdown').selectedItemLabel as 'GB' |
+        'MB';
+  }
+
+  private _initialUnit() {
+    return this._activeDataLimit()?.unit || 'GB';
+  }
+
+  private _initialValue() {
+    return this._activeDataLimit()?.value.toString() || '';
+  }
+
+  private _activeDataLimit(): DisplayDataAmount|undefined {
+    // Returns the limit which currently is enforced on this key, or undefined if there is none.
+    return this.key?.dataLimit || this.serverDefaultLimit;
+  }
+
+  private async _setCustomLimitTapped() {
+    this.showMenu = !this.showMenu;
+    if (this.showMenu) {
+      await this.updateComplete;
+      this._input.focus();
+    }
+  }
+
+  private _setSaveButtonDisabledState() {
```

Comment:
I guess it depends which site you prioritize -- the function definition site or the usage site.  I named it this way so that at the usage site where we hook this in we get more information about what's actually happening than `@keyup=${this._onInputChange}`, which is a little redundant.  WDYT?